### PR TITLE
Updatable Poseidon

### PR DESCRIPTION
**Pull request by 95DDB**

### CODE REVIEW

File: primitives/benches/crypto_primitives/poseidon_crh.rs [hunk 1]
```diff
@@ -8,20 +8,19 @@ use algebra::fields::mnt6753::Fr as MNT6753Fr;
...
 
-    let mut input = Vec::new();
-    input.push(MNT4753Fr::rand(&mut rng));
-    input.push(MNT4753Fr::rand(&mut rng));
-
     c.bench_function("Poseidon CRH Eval for MNT4", move |b| {
         b.iter(|| {
-            MNT4PoseidonHash::evaluate(&input).unwrap();
+            let input = vec![MNT4753Fr::rand(&mut rng); 100];
+            MNT4PoseidonHash::evaluate(input.as_slice()).unwrap();
```
Comment (mkaihara) [inline primitives/benches/crypto_primitives/poseidon_crh.rs:23]: Here, you're evaluating 50 hashes. Perhaps it is more intuitive to measure 100 or 1000 hashes or just one hash.

File: primitives/benches/crypto_primitives/poseidon_crh.rs [hunk 2]
```diff
@@ -30,37 +29,25 @@ fn poseidon_crh_eval_mnt6(c: &mut Criterion) {
...
 
-    let mut input = Vec::new();
-    input.push(MNT6753Fr::rand(&mut rng));
-    input.push(MNT6753Fr::rand(&mut rng));
 
     c.bench_function("Poseidon CRH Eval for MNT6", move |b| {
         b.iter(|| {
-            MNT6PoseidonHash::evaluate(&input).unwrap();
+            let input = vec![MNT6753Fr::rand(&mut rng); 100];
+            MNT6PoseidonHash::evaluate(input.as_slice()).unwrap();
```
Comment (mkaihara) [inline primitives/benches/crypto_primitives/poseidon_crh.rs:36]: The same as MNT4. Perhaps it is more intuitive to measure 100 or 1000 hashes or just one hash.

File: primitives/benches/crypto_primitives/poseidon_crh.rs (hunk 1 above)
Comment (95DDB) [reply primitives/benches/crypto_primitives/poseidon_crh.rs:23]: I tried to reflect the behaviour of the benchmark of normal Poseidon in order to be able to do comparison. Isn't it the same ? Aren't the benches for normal Poseidon also take 100 inputs, i.e. performing 50 permutations (hashes) ?

File: primitives/src/crh/poseidon/updatable/mod.rs [hunk 3]
```diff
@@ -0,0 +1,217 @@
...
+            let padding = personalization.len() % P::R;
+
+            for p in personalization.iter(){
+                instance.update(p);
+            }
+
+            for _ in 0..padding {
+                instance.update(&F::zero());
+            }
+            assert_eq!(instance.pending.len(), 0);
```
Comment (mkaihara) [inline primitives/src/crh/poseidon/updatable/mod.rs:42]: I think you mean: 
if padding != 0 then padding should be P::R - personalisation.len() % P::R

File: primitives/src/crh/poseidon/updatable/mod.rs [hunk 4]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Test initializing UpdatablePoseidonHash with personalization is the same as concatenating
+        // to PoseidonHash input the personalization and the padding.
+        let input = MNT4Fr::rand(&mut rng);
+        let samples = 10;
+        for i in 1..=samples{
+            let personalization = vec![MNT4Fr::rand(&mut rng); i];
+
+            let mut hash_input = personalization.clone();
+            let padding = vec![MNT4Fr::zero(); personalization.len() % MNT4753PoseidonParameters::R];
```
Comment (mkaihara) [inline primitives/src/crh/poseidon/updatable/mod.rs:145]: It should be
let padding;
if personalization.len() % MNT4753PoseidonParameters::R != 0 {
 padding = vec![MNT4Fr::zero(), MNT4753PoseidonParameters::R - personalization.len() % MNT4753PoseidonParameters::R]; }

File: primitives/src/crh/poseidon/updatable/mod.rs [hunk 5]
```diff
@@ -0,0 +1,217 @@
...
+
+        // Test initializing UpdatablePoseidonHash with personalization is the same as concatenating
+        // to PoseidonHash input the personalization and the padding.
+        let input = MNT6Fr::rand(&mut rng);
+        let samples = 10;
+        for i in 1..=samples{
+            let personalization = vec![MNT6Fr::rand(&mut rng); i];
+
+            let mut hash_input = personalization.clone();
+            let padding = vec![MNT6Fr::zero(); personalization.len() % MNT6753PoseidonParameters::R];
```
Comment (mkaihara) [inline primitives/src/crh/poseidon/updatable/mod.rs:201]: The same as MNT4.
It should be 
let padding;
if personalization.len() % MNT6753PoseidonParameters::R != 0 {
 padding = vec![MNT6Fr::zero(),MNT6753PoseidonParameters::R - personalization.len() % MNT6753PoseidonParameters::R]; }

**Review by mkaihara — CHANGES_REQUESTED** [review]
Comments left in the code.

**Pull request closed without merging**